perllib/FixMyStreet/Queue/Item/Report.pm

```diff
@@ -202,39 +219,105 @@ sub _create_reporters {
 
     if (FixMyStreet->staging_flag('send_reports', 0)) {
         # on a staging server send emails to ourselves rather than the bodies
-        %reporters = map { $_ => $reporters{$_} } grep { /FixMyStreet::SendReport::Email/ } keys %reporters;
-        unless (%reporters) {
-            %reporters = ( 'FixMyStreet::SendReport::Email' => FixMyStreet::SendReport::Email->new() );
+        unless (%email_reporters) {
+            %email_reporters
+                = ( EMAIL_SENDER_PREFIX => ${ \EMAIL_SENDER_PREFIX }->new() );
         }
+
+        @other_reporters = ();
     }
 
-    $self->_set_reporters(\%reporters);
+    $self->_set_reporters([@other_reporters, values %email_reporters]);
 }
 
 sub _send {
     my $self = shift;
 
     # Multiply results together, so one success counts as a success.
     my $result = -1;
+    my $report = $self->report;
+
+    my @add_send_fail_body_ids;
+    my @remove_send_fail_body_ids;
+
+    for my $sender ( @{ $self->reporters } ) {
+        my $sender_name = ref $sender;
+        $self->log( 'Sending using ' . $sender_name );
+
+        if ( FixMyStreet->staging_flag( 'send_reports', 0 ) ) {
+            # If on staging with send_reports set to 0, bypass the body_id
+            # logic below, as we may only have a special Email reporter with
+            # no bodies attached
+            my $res = $sender->send( $report, $self->h );
+            $result *= $res;
+            $report->add_send_method($sender_name) if !$res;
+        }
+        elsif ( $sender_name =~ /${\EMAIL_SENDER_PREFIX}/ ) {
```

Comment:
I think this and the else can be combined - as you say in the comment in the else, the senders there will only have one body, so the code for this block will work the same for them too, unless I've missed something.

Reply:
This doesn't work as an email sender can have multiple bodies, but we then add those to the 'To' line and only send the one email. Using the logic in the `else` means multiple emails are sent.

Reply:
I mean, there is otherwise a lot of similarity between the blocks of code so maybe there is a more succinct way of writing it, but this keeps the distinction fairly clear.

Reply:
I meant the opposite of that (which would indeed do as you say) - by "this block" I meant the block I left the comment on. I think you can use the same code from the 'email' part for both it and the else part, ie:

```
if ( FixMyStreet->staging_flag( 'send_reports', 0 ) ) {
    ...
} else {
    my $res = $sender->send( $self->report, $self->h );
    $result *= $res;
    my @body_ids = map { $_->id } @{ $sender->bodies };
    if ($res) {
        push @add_send_fail_body_ids, @body_ids;
    } else {
        $report->add_send_method($sender_name);
        push @remove_send_fail_body_ids, @body_ids;
    }
}
```
There's no need for the else loop, because all the reports that would have gone there only have one body.